src/Symfony/Component/HttpClient/Psr18Client.php

```diff
@@ -0,0 +1,107 @@
+use Psr\Http\Client\RequestExceptionInterface;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * An adapter to turn a Symfony HttpClientInterface into a PSR-18 ClientInterface.
+ *
+ * Run "composer require psr/http-client" to install the base ClientInterface. Run
+ * "composer require nyholm/psr7" to install an efficient implementation of response
+ * and stream factories with flex-provided autowiring aliases.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class Psr18Client implements ClientInterface
+{
+    private $client;
+    private $responseFactory;
+    private $streamFactory;
+
+    public function __construct(HttpClientInterface $client, ResponseFactoryInterface $responseFactory, StreamFactoryInterface $streamFactory)
+    {
+        $this->client = $client;
+        $this->responseFactory = $responseFactory;
+        $this->streamFactory = $streamFactory;
+    }
+
+    public function sendRequest(RequestInterface $request): ResponseInterface
+    {
+        try {
+            $response = $this->client->request($request->getMethod(), (string) $request->getUri(), [
+                'headers' => $request->getHeaders(),
+                'body' => (string) $request->getBody(),
+                'http_version' => '1.0' === $request->getProtocolVersion() ? 1.0 : null,
+            ]);
+
+            $psrResponse = $this->responseFactory->createResponse($response->getStatusCode());
+
+            foreach ($response->getHeaders() as $name => $values) {
+                foreach ($values as $value) {
+                    $psrResponse = $psrResponse->withAddedHeader($name, $value);
+                }
+            }
+
+            return $psrResponse->withBody($this->streamFactory->createStream($response->getContent()));
+        } catch (TransportExceptionInterface $e) {
+            if ($e instanceof \InvalidArgumentException) {
+                throw new Psr18RequestException($e, $request);
+            }
+
+            throw new Psr18NetworkException($e, $request);
+        }
+    }
+}
+
+/**
+ * @internal
+ */
+trait Psr18ExceptionTrait
+{
+    private $request;
+
+    public function __construct(TransportExceptionInterface $e, RequestInterface $request)
+    {
+        parent::__construct($e->getMessage(), 0, $e);
+        $this->request = $request;
+    }
+
+    public function getRequest(): RequestInterface
+    {
+        return $this->request;
+    }
+}
+
+/**
+ * @internal
+ */
+class Psr18NetworkException extends \RuntimeException implements NetworkExceptionInterface
+{
+    use Psr18ExceptionTrait;
+}
+
+/**
+ * @internal
+ */
+class Psr18RequestException extends \InvalidArgumentException implements RequestExceptionInterface
```

Comment:
would `\LogicEcxception` for a base fit better? not sure if an "invalid request" should always be considered an argument value

Reply:
My interpretation maps to using `InvalidArgumentException` here: "the request cannot be sent because it contains invalid state for the use case" => invalid argument provided to the client. I can change of course.

Reply:
> invalid argument provided to the client

right, i think the context of "argument" is arbitrary indeed.

per https://secure.php.net/invalidargumentexception

> Exception thrown if an argument is not of the expected type.

that's any context i guess :D